drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1453 @@
+    PWSK_SOCKET_INTERNAL ListenSocket = l->ListenSocket;
+    PWSK_SOCKET_INTERNAL AcceptSocket = l->AcceptSocket;
+    PWSK_CLIENT_LISTEN_DISPATCH ListenDispatch =
+        (PWSK_CLIENT_LISTEN_DISPATCH)ListenSocket->ListenDispatch;
+
+        /* A PTRANSPORT_ADDRESS address field has an additional
+         * AddressLength field so the struct sockaddr_in starts
+         * at the AddressType (the address family, 2 for AF_INET)
+         * field.
+         */
+    PSOCKADDR RemoteAddress =
+        (PSOCKADDR)(&((PTRANSPORT_ADDRESS)l->ReturnConnectionInfo->RemoteAddress)->Address[0].AddressType);
+    PVOID AcceptSocketContext;
+    const WSK_CLIENT_CONNECTION_DISPATCH *AcceptSocketDispatch;
+    NTSTATUS Status;
+
+    FUNCTION_TRACE;
+
+    if (ListenSocket->CallbackMask & WSK_EVENT_ACCEPT &&
+        ListenDispatch->WskAcceptEvent != NULL &&
+        ListenSocket->ListenIrp != NULL &&
+        !ListenSocket->ListenCancelled)
+    {
+        ListenSocket->ListenIrp = NULL;
+
+        Status = ListenDispatch->WskAcceptEvent(ListenSocket->user_context, 0, &ListenSocket->LocalAddress, RemoteAddress, (PWSK_SOCKET)AcceptSocket, &AcceptSocketContext, &AcceptSocketDispatch);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("ListenDispatch->WskAcceptEvent returned non-successful status 0x%08x\n", Status);
+        }
+        else
+        {
+            memcpy(&AcceptSocket->RemoteAddress, RemoteAddress, sizeof(AcceptSocket->RemoteAddress));
+        }
+
+        /* And wait for the next incoming connection. */
+        /* This is done in a separate thread at IRQL = 0 */
+
+        QueueListening(ListenSocket);
+    }
+    ListenSocket->ListenIrp = NULL;
+
+    SocketPut(AcceptSocket);
+    SocketPut(ListenSocket);
+
+    ExFreePoolWithTag(l->ReturnConnectionInfo, TAG_AFD_TDI_CONNECTION_INFORMATION);
+    ExFreePoolWithTag(l->RequestConnectionInfo, TAG_AFD_TDI_CONNECTION_INFORMATION);
+    ExFreePoolWithTag(l, TAG_NETIO);
+
+    return STATUS_SUCCESS;
+}
+
+static NTSTATUS
+StartListening(_In_ PWSK_SOCKET_INTERNAL ListenSocket)
+{
+    PIRP tdiIrp;
+    NTSTATUS status;
+    struct ListenContext *lc;
+    PWSK_SOCKET_INTERNAL AcceptSocket;
+
+    FUNCTION_TRACE;
+
+    if (ListenSocket->LocalAddressHandle == NULL)
+    {
+        DPRINT1("LocalAddressHandle is not set, need to bind your socket before listening\n");
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    status = CreateSocket(ListenSocket->family, ListenSocket->type, ListenSocket->proto, WSK_FLAG_CONNECTION_SOCKET, &AcceptSocket);
+    if (status != STATUS_SUCCESS)
+    {
+        DPRINT1("Could not create AcceptSocket, status is 0x%08x\n", status);
+        return status;
+    }
+    AcceptSocket->ListenSocket = ListenSocket;
+    /* Put when the AcceptSocket is closed */
+    SocketGet(AcceptSocket->ListenSocket);
+
+    status = STATUS_INSUFFICIENT_RESOURCES;
+    lc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*lc), TAG_NETIO);
```

Comment:
This gives better semantics.
```suggestion
    lc = ExAllocatePoolUninitialized(NonPagedPool, sizeof(*lc), TAG_NETIO);
```

Reply:
Ok applied, thanks